Federation in the platform (#504)
apollographql/apollo: docs/source/platform/federation.mdx

## Managed Federation
[//]: # (Description: This section should discuss the basic idea of managed federation without getting into specific and talk vaguely of the problem of service rollout and looking for a balance of automation and observability, workflow, etc.)
[//]: # (Assignee: Jackson)
[//]: # (Reviewer: Adam)

In small scale cases it can be sufficient to define the list of services comprising a federated graph directly within the gateway, and manually update the list and restart the server as services are modified and created. However for larger scale projects it is helpful to be able to remotely manage the gateway, such that new services may be added, or existing services modified, all with zero downtime.

For these use cases, a gateway can be configured to pull a remotely hosted service list from the [Schema Registry](https://www.apollographql.com/docs/platform/schema-registry/) on `Apollo Dashboard`. This is `Managed Federation`, and it is a key component of building a robust federated graph that can be concurrently developed by teams of any size.

#### Overview

> // TODO: Document the overview of managing federation workflow, happy path

Reviewer: @zionts 	Is this unassigned section removable? The content seems to be in the "Managed Federation" section already?

---

> title: Federation in Production

Reviewer: rename title to "Federation on the Platform"

> description: How to run with Federation in production, leveraging the Apollo Platform and focusing on reliability and observability

Reviewer: TODO: intro too long

The key difference between running GraphQL with a monolithic server instead of running multiple services orchestrated by a [gateway](https://www.apollographql.com/docs/apollo-server/federation/implementing/#running-a-gateway) is that your GraphQL schema is dynamically defined by composing an overall schema from the [partial schemas](<LINK_TO_GLOSSARY>) of GraphQL services that comply with the [federation spec](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/). This empowers teams to own a portion of the graph, and move faster by allowing changes to be independently developed and deployed.

The Apollo GraphQL platform provides tools to help ensure that this increase in autonomy doesn't come at a cost to stability. Since the GraphQL schema is the central hub of the data in your graph, it's important to [validate](<LINK_TO_SCHEMA_VALIDATION>) that any changes to a particular service's partial schema won't cause breakages in other parts of your graph, or affect existing [operations made by clients](LINK_TO_METRICS), so you can scale your data graph with confidence.

> ## Inspecting your Graph

Reviewer: Move "regsitering" before "inspecting"

Reviews    http://localhost:4002/graphql  3 July 2019 (2 days ago)

View full details at: https://engine.apollographql.com/graph/service-list-federation-demo/service-list
```

Navigate to the Services tab in the Apollo Platform to see an interactive list of information about the implementing services in your graph.

<div style="text-align:center">
    <img src="../images/ServicesListPlatform.png" alt="Service List">
    </img>
</div>

[//]: # (Description: How to see the current services in your graph. This section might be expanded later to discuss showing history, composition results, etc. This section should include some screenshots)
[//]: # (Assignee: Maya)
[//]: # (Reviewer: Justin)

## Registering your graph
> Using the `apollo service:push` and `apollo service:check` command to track changes to your graph

[//]: # (Description: An overview of how and when to use `apollo service:push` and `apollo service:check` in your CI/CD workflow, linking out to the specifics of usage in the specific docs (#502 and #503). Ideally this section includes some screenshots of the workflows in practice.)
[//]: # (Assignee: James)
[//]: # (Reviewer: Adam Z)

> ## Running with managed federation

Reviewer: justify why

There is a wealth of information around reliability in a distributed service-oriented architecture, and those best practices still certainly apply here. For instance, you should ensure that the resource needs of your services are well-understood, that services and the gateway only roll over to a new version after passing health checks, and you should strive to make all services stateful when possible. This document is meant for teams with the need for a federated GraphQL layer, to provide details of nuance in how federation works, how the gateway responds to changes, and best practices in operating federation in production. If you're running with federation in production, we'd love to [hear from you](mailto:federation@apollographql.com) and accept any [contribution](https://github.com/apollographql/apollo-server/blob/master/docs/source/federation/production.md) to this documentation.

## Best Practices

> [//]: # (Description: This section should basically introduce that talking with people running gateways in production (and running it ourselves), we've collected some best practices to share)

Reviewer: this is not hidden

[//]: # (Assignee: Adam)
[//]: # (Reviewer: Pierre, James)

In operating federation in production ourselves and working with a variety of teams deploying federation in their environments, we have collected some best practices to maintain reliability and control over a federated GraphQL layer at scale. If you're running federation in your infrastructure, we'd love to [hear from you](mailto:federation@apollographql.com) to help share any best practices you and your team may have learned from operating federation at scale.

> #### Treat the gateway as infrastructure

Reviewer: larger heading

For instance, in order to have a canary deployment, you might maintain two production graphs in the [Schema Registry](https://www.apollographql.com/docs/platform/schema-registry), one called `prod` and one called `prod-canary`. Your deployment of a change to some implementing service named "foo" might look something like this:

> 1. Check changes in "foo" against `prod` and `prod-canary`: `apollo service:check --tag=prod --serviceName=foo && apollo service:check --tag=prod-canary --serviceName=foo`

Reviewer: command too long

1. Wait for health checks to pass against the canary, watch dashboards, etc.
1. After the canary is stable, roll out the changes to the rest of production using `apollo service:push --tag=prod --serviceName=foo`

Because you can [tag metrics with variants](https://www.apollographql.com/docs/platform/schema-registry/#associating-metrics-with-a-variant) as well, you can use [Apollo Graph Manager](https://engine.apollographql.com) to verify a canary's performance before rolling out changes to the rest of the graph. You can also use a similar strategy with variants to support a variety of other advanced deployment workflows, like blue/green deployments.

## Managed Federation
[//]: # (Description: This section should discuss the basic idea of managed federation without getting into specific and talk vaguely of the problem of service rollout and looking for a balance of automation and observability, workflow, etc.)
[//]: # (Assignee: Jackson)
[//]: # (Reviewer: Adam)

> In small scale cases it can be sufficient to define the list of services comprising a federated graph directly within the gateway, and manually update the list and restart the server as services are modified and created. However for larger scale projects it is helpful to be able to remotely manage the gateway, such that new services may be added, or existing services modified, all with zero downtime.

Reviewer: move this to "running with federation"

#### Reliability

[//]: # (Description: This section should document how the gateway polls GCS for updates, why it's a reliable model, what the defaults are, and any recommendations)
[//]: # (Assignee: Jackson)
[//]: # (Reviewer: Adam)

When operating as a managed gateway, `Apollo Server` will poll the [Schema Registry](https://www.apollographql.com/docs/platform/schema-registry/) for updates to the registered service list at 10 second intervals. When a service list update does occur, `Apollo Server` will create a new composed schema from the federated services, and begin to roll over to serving the new schema. Existing in-flight operations on the old schema will continue to be processed while serving the new schema. For this reason, it can be helpful to serve the new schemas from new endpoints, such that no downtime is incurred during rollover.

In the event a network failure prevents an `Apollo Server` gateway from contacting the Schema Registry, the gateway will continue to serve the last known schema while it attempts to reestablish a connection to the registry.

> // TODO: Document the model of operating on top of GCS, ways of falling back to local files & introspection

Reviewer: remove, outdated

The Apollo GraphQL platform provides tools to help ensure that this increase in autonomy doesn't come at a cost to stability. Since the GraphQL schema is the central hub of the data in your graph, it's important to [validate](<LINK_TO_SCHEMA_VALIDATION>) that any changes to a particular service's partial schema won't cause breakages in other parts of your graph, or affect existing [operations made by clients](LINK_TO_METRICS), so you can scale your data graph with confidence.

## Registering your graph

If you're running a distributed GraphQL infrastructure, where implementing services [compose](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/#fetch-service-capabilities) to form a complete schema, there's functionality to register each service's partial schema and still get all of the benefits of the schema registry.  

> To push a single implementing service to Apollo, run `apollo service:push` with the `--serviceName` and `--endpoint` tags. The CLI will know where to fetch your local partial schema based on this endpoint. Note that running `apollo service:push` on any implementing service also registers the overall schema of the graph and writes files to GCS.

Author: EXAMPLE of what `service:push` output looks like when it fails to compose, but still updates the implementing service

```
$ npx apollo@2.15.1-alpha.1 service:push --serviceName registry --serviceURL http://engine-apigateway:4001/api/graphql --tag staging --endpoint http://localhost:4001/api/graphql --engine https://engine-staging-graphql.apollographql.com/api/graphql
  ✔ Loading Apollo Project
  ✔ Loading Apollo Project
  ✔ Uploading service to Engine

The 'registry' service for the 'apollo-federated@staging' graph was updated

*THE SERVICE UPDATE RESULTED IN COMPOSITION ERRORS.*

Composition errors must be resolved before the graph's schema or corresponding gateway can be updated.
For more information, see https://www.apollographql.com/docs/apollo-server/federation/errors/

Error   [registry] ServiceMutation.upsertImplementingService -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.upsertImplementingServiceAndTriggerComposition -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.uploadPartialSchema -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.validatePartialSchemaOfImplementingServiceAgainstGraph -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.recalculateCompositionConfig -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.removeImplementingServiceAndTriggerComposition -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.removeImplementingService -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.checkPartialSchema -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.registerOperations -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.registerOperationsWithResponse -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.uploadSchema -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.deleteSchema -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.deleteInvalidSchema -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.checkSchema -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.tagSchema -> requires the field `service` to be marked as @external.
Error   [registry] ServiceMutation.deleteSchemaTag -> requires the field `service` to be marked as @external.

The gateway for the 'apollo-federated@staging' graph was NOT updated with a new schema

```

[//]: # (Description: An overview of the concept of managed federation, linking out to the details on gateway defaults w/ an API key and the specfics on setting up the lifecycle options. Ideally this section has a diagram of managed federation lifecycle.)
[//]: # (Assignee: Adam Z)
[//]: # (Reviewer: Trevor)

In small scale cases it can be sufficient to define the list of services comprising a federated graph directly within the gateway, and manually update the list and restart the server as services are modified and created. However for larger scale projects it is helpful to be able to remotely manage the gateway, such that new services may be added, or existing services modified, all with zero downtime.

For these use cases, a gateway can be configured to pull a remotely hosted service list from the [Schema Registry](https://www.apollographql.com/docs/platform/schema-registry/) on `Apollo Dashboard`. This is `Managed Federation`, and it is a key component of building a robust federated graph that can be concurrently developed by teams of any size.

# Running Federation in Production
> With a focus on reliability and robustness

> Like any distributed architecture, you should make sure that your federated GraphQL layer has proper observability, monitoring, and automation to ensure reliability and performance of both your gateway and the federated services underneath it. Serving your GraphQL API from a distributed architecture has many benefits, like productivity, isolation, and being able to match the right services with the right runtimes. Operating a distributed system also has more complexity and points of failure than operating a monolith, and with that complexity comes a need to heighten observability into the state of your system and control over its coordination.

Reviewer 2: we can say this instead

Like any distributed architecture, definitely make sure that a federated GraphQL layer has proper observability, monitoring, and automation to ensure reliability and performance of both the gateway and the federated services underneath it. Serving your GraphQL API from a distributed architecture has many benefits, like productivity, isolation, and being able to match the right services with the right runtimes. Operating a distributed system also has more complexity and points of failure than operating a monolith, and with that complexity comes a need to heighten observability into the state of your system and control over its coordination.

For instance, in order to have a canary deployment, you might maintain two production graphs in the [Schema Registry](https://www.apollographql.com/docs/platform/schema-registry), one called `prod` and one called `prod-canary`. Your deployment of a change to some implementing service named "foo" might look something like this:

1. Check changes in "foo" against `prod` and `prod-canary`:  
   ```

>    apollo service:check --tag=prod --serviceName=foo && apollo service:check --tag=prod-canary --serviceName=foo

Reviewer 3: For readability, can we place the second command on a newline?

   ```
   (_Note: If composition fails due to intermediate changes to the canary graph, new configuration will not be rolled out_)
1. Wait for health checks to pass against the canary, watch dashboards, etc.
1. After the canary is stable, roll out the changes to the rest of production using
  ```
  apollo service:push --tag=prod --serviceName=foo
  ```

Because you can [tag metrics with variants](https://www.apollographql.com/docs/platform/schema-registry/#associating-metrics-with-a-variant) as well, you can use [Apollo Graph Manager](https://engine.apollographql.com) to verify a canary's performance before rolling out changes to the rest of the graph. You can also use a similar strategy with variants to support a variety of other advanced deployment workflows, like blue/green deployments.

### Controlling rollout

To begin, ensure all component services are up and running, and the working directory has a `.env` file containing the appropriate `ENGINE_API_KEY` for your federated service. Then, upload each service to the gateway through `Apollo CLI`:

```bash
apollo service:push       \
  --serviceURL={{url}}    \
  --endpoint={{url}}      \
  --serviceName={{name}}  \

>   --federated

Author: I'm pretty sure we don't need `--federated` anymore, right?

```

> Each service needs a unique name, this is how you will identify the service for schema updates

The given `endpoint` will be queried for the new schema which, if compatible with the rest of the schema, will be uploaded to `Apollo Graph Manager`. At this point, any gateways serving this graph will begin to roll over to the new schema.

> You can later modify hosted service definitions by rerunning the `apollo service:push` command with the service's name and the new URL.

Author: why would the url be new?

Author: maybe new partial schema?

> > What's the difference between `serviceURL` and `endpoint` parameters? The `endpoint` parameter controls the endpoint where the schema will be fetched from at composition, whereas `serviceURL` controls what URL the gateway will query at runtime. This is especially helpful if you have disabled introspection queries in your production services.

Author: I don't understand the last sentence here, but I love that we're explaining the difference between these two similarly named options.

// TODO: Document the integration with service:push, services under a graph, and gateway lifecycle management utilities

### Reliability

[//]: # (Description: This section should document how the gateway polls GCS for updates, why it's a reliable model, what the defaults are, and any recommendations)
[//]: # (Assignee: Jackson)
[//]: # (Reviewer: Adam)

> When operating as a managed gateway, `Apollo Server` will poll the [Schema Registry](https://www.apollographql.com/docs/platform/schema-registry/) for updates to the registered service list at 10 second intervals. When a service list update does occur, `Apollo Server` will create a new composed schema from the federated services, and begin to roll over to serving the new schema. Existing in-flight operations on the old schema will continue to be processed while serving the new schema. For this reason, it can be helpful to serve the new schemas from new endpoints, such that no downtime is incurred during rollover.

Author: I think we should have additional information here about the fact that this isn't polling the schema registry directly, but rather a cloud file system over a CDN. The whole point is that the reliability of our infrastructure isn't critical for the reliability of your graph.

> In the event a network failure prevents an `Apollo Server` gateway from contacting the Schema Registry, the gateway will continue to serve the last known schema while it attempts to reestablish a connection to the registry.

Author: That's all well and good for currently running gateways, but what about a restart? There should be the ability for the user to define a fallback mechanism.

<!--

>   Jackson: There is no way of falling back to local files/introspection currently, but I also don't think there should be, consider:

Author: I think there should certainly be a way of falling back to local files. GCS being down shouldn't mean that the entirety of your graph is down.

---
title: Federation
description: Leveraging the Apollo Graph Manager for Federation
---
[//]: # (Description: Introductory content, discussing the overall model of the federation platform, why you would want to register partial schemas, and framing the concept of thinking about running federation as a coordination and collaboration problem to motivate the tools we've built)
[//]: # (Assignee: Chang)
[//]: # (Reviewer: Caydie, James)

> Organizations expose their data as a single graph in order to have a complete view of all of their organizational data along with well-defined and reusable patterns for data access, observability, security, caching, and other API-level considerations. However, even though exposing all of this data in a unified way has [huge benefits](/intro/benefits/), it comes with additional coordination across multiple teams, all of whom own a portion of the data graph. As an organization, ideally individual teams are able to work independently on GraphQL microservices without needing to coordinate changes with other teams. By the same token, you want to ensure that changes from each team are safe for the entirety of the graph and don't violate any assumptions or cause conflicts with other service teams.

Author: We need to mention the concept of the Apollo Graph Manager here if we're going to follow it up with a section on that manager

### At a glance

The Graph Manager maintains both the [current state](https://www.apollographql.com/docs/platform/schema-registry/#using-the-schema-registry) of your graph as well as the [history](https://www.apollographql.com/docs/platform/schema-registry/#viewing-schema-change-history) of changes to that graph. When running GraphQL as a monolithic service with a static schema, the state of a data graph is well-represented by two components: operations and the schema. The schema represents what data is available in the graph, and operations, whether [registered](https://www.apollographql.com/docs/platform/operation-registry/#3-register-operations-from-your-client-bundle) or [collected dynamically](https://www.apollographql.com/docs/apollo-server/features/metrics/), represent the usage of that data graph.

![Graph State](../img/federation/graph-current-state.png)

The concept of a Federated Graph adds a single but important piece into the model for running and coordinating over a Graph: **One graph is composed of multiple implementing services**. This means that the state of the graph includes the set of services that implement it as well as _their_ states.

> The core concept of the Graph Manager in supporting federation is that each implementing service's partial schema should be tracked the same way as a monolithic data graph would be, and that the state of the Graph should be a reflection of the state of all of its dependencies.

Author: split into two sentences

> ## Registering your implementing services

Author: There should be a sub-section in here about using `service:delete`

If you're running a distributed GraphQL infrastructure, where implementing services [compose](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/#fetch-service-capabilities) to form a complete schema, there's functionality to register each service's partial schema and still get all of the benefits of the schema registry.

> To push a single implementing service to Apollo, run `apollo service:push` with the `--serviceName` and `--endpoint` tags. The CLI will know where to fetch your local partial schema based on this endpoint. Note that running `apollo service:push` on any implementing service also registers the overall schema of the graph and writes files to GCS.

Author: you also always need a `--serviceURL`
We should also mention the idea of registering to variants, linking out to the doc on managing environments, using the `--tag` flag (though ideally we could ship a change for the `--variant` flag.

Author: I think we should additionally point out both here **AND** in the managed federation (deployment? gateway? idk, let's be consistent and we can rename en masse later) section that the `serviceURL` is used for the gateway coordinating traffic, so it should be an URL that is accessible for the gateway to access the implementing services.

> To push a single implementing service to Apollo, run `apollo service:push` with the `--serviceName` and `--endpoint` tags. The CLI will know where to fetch your local partial schema based on this endpoint. Note that running `apollo service:push` on any implementing service also registers the overall schema of the graph and writes files to GCS.

Author: We should repeat the instructions of having an API key in the `.env` file of your working directory for this to work

> To push a single implementing service to Apollo, run `apollo service:push` with the `--serviceName` and `--endpoint` tags. The CLI will know where to fetch your local partial schema based on this endpoint. Note that running `apollo service:push` on any implementing service also registers the overall schema of the graph and writes files to GCS.

Author: We need to say early on which version of `apollo` to be using, let's start at `>=2.15.1`

As an example, running service push on a local “Accounts” implementing service might look like:

```

> ~$apollo service:push --serviceName="Accounts" --serviceURL="http://accounts-graphql:4001/ --endpoint="http://localhost:4001/"

Author: add closing quotation

  ✔ Loading Apollo Project
  ✔ Uploading service to Engine
id      schema                        tag
──────  ────────────────────────────  ───────
190331  service-list-federation-demo  current
```

> Pushing an implementing service to Apollo will update the [Services tab](#inspecting-your-graph).

Author: We should add a `>NOTE` or an additional sub-section here around handling failures in composition and how the implementing service will always be updated as long as its partial schema is "valid," but the graph will **not** be updated (i.e. the schema and the managed gateway config (also sidenote: let's align on the term "managed gateway config" instead of saying weird stuff about GCS everywhere))

Author: This also points out that if we're going to mention managed config here, we either need to bring that section higher in the document or just link downwards to it.

Author: We should also show an example output of what happens when service:push fails, and update this to have the new content that was added today in the messages

Author: EXAMPLE of what happens when you push a service and it composes fine and updates files. We should modify names and use the same example of services & graph across the board

GRAPH: space-explorer
VARIANT: current
SERVICES: users, launches, reservations

```
$ ../apollo-tooling/packages/apollo/bin/run service:push --serviceName registry --endpoint http://localhost:4001/api/graphql --engine https://engine-staging-graphql.apollographql.com/api/graphql --serviceURL http://engine-apigateway.svc.cluster.local:4001/api/graphql --tag staging
  ✔ Loading Apollo Project
  ✔ Loading Apollo Project
  ✔ Uploading service to Engine

The 'registry' service for the 'apollo-federated@staging' graph was updated

The gateway for the 'apollo-federated@staging' graph was updated with a new schema, composed from the updated 'registry' service

id      graph             variant
──────  ────────────────  ───────
da81e5  apollo-federated  staging

```

> 

Author: Before diving into the Inspecting your graph, give a sub-section with a sneak peek of managed federation here, noting that you don't want the reliability of your gateway to depend on the reliability of your services, so do not use introspection, use another source of truth. There is a default implementation of this that integrates with the registry and is shipped by default in the Apollo Gateway.

```
~$ apollo service:list
  ✔ Loading Apollo Project
  ✔ Fetching list of services for graph service-list-federation-demo

name       URL                            last updated
─────────  ─────────────────────────────  ────────────────────────
Accounts   http://localhost:4001/graphql  3 July 2019 (2 days ago)
Inventory  http://localhost:4004/graphql  3 July 2019 (2 days ago)
Products   http://localhost:4003/graphql  3 July 2019 (2 days ago)
Reviews    http://localhost:4002/graphql  3 July 2019 (2 days ago)

View full details at: https://engine.apollographql.com/graph/service-list-federation-demo/service-list
```

Navigate to the Services tab in the Apollo Platform to see an interactive list of information about the implementing services in your graph.

<div style="text-align:center">
    <img src="../images/ServicesListPlatform.png" alt="Service List">
    </img>
</div>

[//]: # (Description: How to see the current services in your graph. This section might be expanded later to discuss showing history, composition results, etc. This section should include some screenshots)
[//]: # (Assignee: Maya)
[//]: # (Reviewer: Justin)

> ## Validating changes to your graph

Author: Pasting some examples here so that we can use them in the section.

Author: EXAMPLE of what `apollo service:check` looks like with SAFE changes:
```
$ npx apollo@2.15.1-alpha.1 service:check --serviceName registry --endpoint http://localhost:4001/api/graphql --engine https://engine-staging-graphql.apollographql.com/api/graphql --tag staging --frontend https://engine-staging.apollographql.com
  ✔ Loading Apollo Project
  ✔ Found 0 graph composition errors for service registry on graph apollo-federated
  ✔ Validated composed schema against tag staging on graph apollo-federated
  ✔ Compared 2 schema changes against 56 operations over the last 548 days
  ✔ Found 0 breaking changes and 2 compatible changes

PASS    FIELD_ADDED  `ServiceMutation.createApiKeyAndWriteStorageSecret` was added
PASS    FIELD_ADDED  `ServiceMutation.uploadSchema` was added

View full details at: https://engine-staging.apollographql.com/service/apollo-federated/check/5f209250-000e-45f4-9d5c-ff022225f542
```

Author: EXAMPLE of what `apollo service:check` looks like with UNSAFE changes:
```
$ ../apollo-tooling/packages/apollo/bin/run service:check --serviceName registry --endpoint http://localhost:4001/api/graphql --engine https://engine-staging-graphql.apollographql.com/api/graphql --tag staging --frontend https://engine-staging.apollographql.com
  ✔ Loading Apollo Project
  ✔ Found 0 graph composition errors for service registry on graph apollo-federated
  ✔ Validated composed schema against tag staging on graph apollo-federated
  ✔ Compared 1 schema change against 55 operations over the last 548 days
  ✖ Found 1 breaking change and 0 compatible changes
    → breaking changes found

FAIL    FIELD_REMOVED  `ServiceMutation.uploadSchema` was removed

View full details at: https://engine-staging.apollographql.com/service/apollo-federated/check/941c4fa2-c716-4df4-9fc5-130b88c8ebca
```

Reviewer 2: Add a note that validation does not update the graph

Author: EXAMPLE of what `apollo service:check` looks like with composition errors:
```
$ ../apollo-tooling/packages/apollo/bin/run service:check --serviceName registry --endpoint http://localhost:4001/api/graphql --engine https://engine-staging-graphql.apollographql.com/api/graphql --tag staging
  ✔ Loading Apollo Project
  ✖ Found 16 graph composition errors for service registry on graph apollo-federated
    → Federated service composition was unsuccessful. Please see the reasons below.
    Validating composed schema against tag staging on graph apollo-federated
    Comparing schema changes
    Reporting result

Service   Field                                                                   Message
────────  ──────────────────────────────────────────────────────────────────────  ───────────────────────────────────────────────────────
registry  ServiceMutation.upsertImplementingService                               requires the field `service` to be marked as @external.
registry  ServiceMutation.upsertImplementingServiceAndTriggerComposition          requires the field `service` to be marked as @external.
registry  ServiceMutation.uploadPartialSchema                                     requires the field `service` to be marked as @external.
registry  ServiceMutation.validatePartialSchemaOfImplementingServiceAgainstGraph  requires the field `service` to be marked as @external.
registry  ServiceMutation.recalculateCompositionConfig                            requires the field `service` to be marked as @external.
registry  ServiceMutation.removeImplementingServiceAndTriggerComposition          requires the field `service` to be marked as @external.
registry  ServiceMutation.removeImplementingService                               requires the field `service` to be marked as @external.
registry  ServiceMutation.checkPartialSchema                                      requires the field `service` to be marked as @external.
registry  ServiceMutation.registerOperations                                      requires the field `service` to be marked as @external.
registry  ServiceMutation.registerOperationsWithResponse                          requires the field `service` to be marked as @external.
registry  ServiceMutation.uploadSchema                                            requires the field `service` to be marked as @external.
registry  ServiceMutation.deleteSchema                                            requires the field `service` to be marked as @external.
registry  ServiceMutation.deleteInvalidSchema                                     requires the field `service` to be marked as @external.
registry  ServiceMutation.checkSchema                                             requires the field `service` to be marked as @external.
registry  ServiceMutation.tagSchema                                               requires the field `service` to be marked as @external.
registry  ServiceMutation.deleteSchemaTag                                         requires the field `service` to be marked as @external.
```
We should obviously pare these composition errors down.

With a federated graph, use the `apollo service:check` command to validate individual service changes by adding the `--serviceName` flag. Engine will run composition on the proposed partial schema with the current list of implementing services and their partial schemas, making sure that the composition is successful. That composed schema will then be diff'ed against the most recently registered schema and validate that those changes are safe.

> Using the `apollo service:push` and `apollo service:check` command to track changes to your graph

<!--
[//]: # (Description: An overview of how and when to use `apollo service:push` and `apollo service:check` in your CI/CD workflow, linking out to the specifics of usage in the specific docs (#502 and #503). Ideally this section includes some screenshots of the workflows in practice.)
[//]: # (Assignee: James)
[//]: # (Reviewer: Adam Z)
-->

> ## Running with managed federation

Author: Rename to Managed Federation

This section should bring in all of the details from the below managed federation section and likely have some sub-sections, with an introduction.

> Taking control of how service changes bubble up to the Apollo Gateway

[//]: # (Description: An overview of the concept of managed federation, linking out to the details on gateway defaults w/ an API key and the specfics on setting up the lifecycle options. Ideally this section has a diagram of managed federation lifecycle.)
[//]: # (Assignee: Adam Z)
[//]: # (Reviewer: Trevor)

In small scale cases it can be sufficient to define the list of services comprising a federated graph directly within the gateway, and manually update the list and restart the server as services are modified and created. However for larger scale projects it is helpful to be able to remotely manage the gateway, such that new services may be added, or existing services modified, all with zero downtime.

For these use cases, a gateway can be configured to pull a remotely hosted service list from the [Schema Registry][] on `Apollo Dashboard`. This is `Managed Federation`, and it is a key component of building a robust federated graph that can be concurrently developed by teams of any size.

> # Running Federation in Production

Author: take this introduction, and re-purpose it for the `Managed Federation` section that comes after `validating your schema`

The below section of "Best Practices" should remain as is and stand alone.

> With a focus on reliability and robustness

Like any distributed architecture, you should make sure that your federated GraphQL layer has proper observability, monitoring, and automation to ensure reliability and performance of both your gateway and the federated services underneath it. Serving your GraphQL API from a distributed architecture has many benefits, like productivity, isolation, and being able to match the right services with the right runtimes. Operating a distributed system also has more complexity and points of failure than operating a monolith, and with that complexity comes a need to heighten observability into the state of your system and control over its coordination.

> There is a wealth of information around reliability in a distributed service-oriented architecture, and those best practices still certainly apply here. For instance, you should ensure that the resource needs of your services are well-understood, that services and the gateway only roll over to a new version after passing health checks, and you should strive to make all services stateful when possible. This document is meant for teams with the need for a federated GraphQL layer, to provide details of nuance in how federation works, how the gateway responds to changes, and best practices in operating federation in production. If you're running with federation in production, we'd love to [hear from you](mailto:federation@apollographql.com) and accept any [contribution](https://github.com/apollographql/apollo-server/blob/master/docs/source/federation/production.md) to this documentation.

Reviewer 2: this might be better:

There is a wealth of information around reliability in a distributed service-oriented architecture, and those best practices still certainly apply here. For instance, ensure that the resource needs of services are well-understood, that services and the gateway only roll over to a new version after passing health checks, strive to make all services stateful when possible. This document is meant for teams with the need for a federated GraphQL layer, to provide details of nuance in how federation works, how the gateway responds to changes, and best practices in operating federation in production. If you're running with federation in production, we'd love to [hear from you](mailto:federation@apollographql.com) and accept any [contribution](https://github.com/apollographql/apollo-server/blob/master/docs/source/federation/production.md) to this documentation.

```bash
apollo service:push       \
  --serviceURL={{url}}    \
  --endpoint={{url}}      \
  --serviceName={{name}}  \
  --federated
```

> Each service needs a unique name, this is how you will identify the service for schema updates

The given `endpoint` will be queried for the new schema which, if compatible with the rest of the schema, will be uploaded to `Apollo Graph Manager`. At this point, any gateways serving this graph will begin to roll over to the new schema.

You can later modify hosted service definitions by rerunning the `apollo service:push` command with the service's name and the new URL.

> What's the difference between `serviceURL` and `endpoint` parameters? The `endpoint` parameter controls the endpoint where the schema will be fetched from at composition, whereas `serviceURL` controls what URL the gateway will query at runtime. This is especially helpful if you have disabled introspection queries in your production services.

// TODO: Document the integration with service:push, services under a graph, and gateway lifecycle management utilities

> ### Reliability

Author: This section should go into the managed federation section

[Schema Registry]: /platform/schema-registry/

In the event a network failure prevents an `Apollo Server` gateway from contacting the Schema Registry, the gateway will continue to serve the last known schema while it attempts to reestablish a connection to the registry.

<!--
  Jackson: There is no way of falling back to local files/introspection currently, but I also don't think there should be, consider:
  The usage model of managed gateways is:
  1. Launch services
  2. Post their schema's to registry
  3. Update services with new schemas and URLs
  4. GOTO 2
  If a network failure occurs after several updates and the gateway falls back to a locally stored list, that list will almost certainly be outdated. On the other hand, if it falls back to the last known good configuration, it has a much higher chance of still being active.
-->

> ### Enabling Federated Metrics

Author: We should add a new top-level section (after Best Practices, same number of lil pound things) called "Observability" and put federated metrics and inspecting query plans within it.

IF I REMEMBER CORRECTLY, we can literally take one of the paragraphs from the intro to `RUNNING IN PRODUCTION` for observability and the other for managed federation.

---
title: Managing a federated graph
description: How to run, manage, and deploy a federated graph
---

When organizations roll out GraphQL across multiple teams, coordinating and maintaining a single monolithic schema becomes difficult and error-prone. [Apollo Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/) solves this problem by enabling teams to build a distributed graph composed of multiple services declaratively without fragile stitching code.

> While federation empowers teams to scale GraphQL much faster than before, it also introduces complexity that comes along with running a distributed system. Teams need federation-aware tooling in order to help them coordinate updating the graph whenever an underlying service changes. Apollo's **graph manager** is a turnkey solution to this problem that includes federated service checks, managed service deployments, and federation-aware analytics out of the box for teams on a paid plan.

Author: do we want to remove the paid plan portion here since all of this is free now?

To push a single implementing service to Apollo, run `apollo service:push` with the `--serviceName`, `--endpoint`, and `--serviceURL` tags. The CLI will know where to fetch your local partial schema based on the `--endpoint` flag, and the `--serviceURL` flag indicates where the implementing service can be reached by the gateway. The `--serviceName` flag is used as a unique identifier for each implementing service.

Make sure you have a `.env` file locally with your ENGINE_API_KEY defined! To get an API key, click [here](https://engine.apollographql.com). To create a new `.env` file, copy your API key into the following command from your terminal:
```
echo "ENGINE_API_KEY={{service:foo:asd9fjasdor8her}}">>.env
```

As an example, running service push on a “launches” implementing service might look like:

```
~$ apollo service:push --serviceName="launches" \
                      --serviceURL="http://launches-graphql.svc.cluster.local:4001/" \
                      --endpoint="http://localhost:4001/"
  ✔ Loading Apollo Project
  ✔ Uploading service to Engine

> The 'registry' service for the 'space-explorer@current' graph was updated

Reviewer 2: this should say 'launches' service not 'registry' service D:

    Validating composed schema against tag current on graph space-explorer
    Comparing schema changes
    Reporting result

Service   Field                                                                   Message
────────  ──────────────────────────────────────────────────────────────────────  ───────────────────────────────────────────────────────
launches  Mutation.scheduleLaunch                                                       requires the field `service` to be marked as @external.
```

#### Failed usage checks
```
$ apollo service:check --serviceName=launches --endpoint=http://localhost:4001/api/graphql
  ✔ Loading Apollo Project
  ✔ Found 0 graph composition errors for service registry on graph space-explorer
  ✔ Validated composed schema against tag current on graph space-explorer
  ✔ Compared 1 schema change against 55 operations over the last 548 days
  ✖ Found 1 breaking change and 0 compatible changes
    → breaking changes found

FAIL    FIELD_REMOVED  `Mutation.scheduleLaunch` was removed

View full details at: https://engine.apollographql.com/service/space-explorer/check/941c4fa2-c716-4df4-9fc5-130b88c8ebca
```

> ## Managed federation

Reviewer 2: we should rename this managed configuration - sorry for the flip flopping >.<

1. The partial schema is uploaded to a secure location and indexed
1. The implementing service is updated in the registry to point to the partial schema
1. All implementing services are composed in the cloud to produce a new complete schema
1. If composition fails, the command exits and emits errors
1. If composition succeeds, the top-level managed configuration file is updated in-place to point to the updated set of implementing services

On the other side of the equation, sits the gateway. The gateway is constantly listening for changes to the top-level managed configuration file. The location of the managed configuration file is guarded by using a hash of the API key, provisioned ahead of time so as not to affect reliability. The life-cycle of dynamic configuration updates is as follows:

> 1. The gateway listens for updates to its managed configuration every 10 seconds

Reviewer 3: Every 10s is an implementation detail that will be configurable and possibly changed down the road, I'll suggest staying slightly less specific here.
```suggestion
1. The gateway listens for updates to its managed configuration
```

1. On update, the gateway downloads configuration for each implementing service in parallel

> 1. The gateway updates its query-planner to use the new configuration

Reviewer 3: This feels redundant, or collapsible into the item above. Query planner doesn't need to be hyphenated.

> 1. The gateway flushes its `queryPlan` cache to begin planning new queries with the new config

Reviewer 3: This is an implementation detail of the gateway that can probably be left out (at least the cache bit). We could say:
>The gateway begins using its new configuration to resolve incoming requests, while in-flight requests are resolved with the previous configuration

>The gateway continues to resolve in-flight requests with the previous configuration while using the updated configuration for all new requests

I prefer the 2nd - 1st part is the interesting stuff, 2nd part seems like a given.

> 1. In-flight requests finish being served and delivered to consumers as incoming operations are served as well

Reviewer 3: Potentially collapsed into the suggestion above

> 

Reviewer 3: We should mention when composition occurs in the list. I think that's a particularly interesting part of the sequence of events.

### Integrating with CI / CD

### Reliability and security

> The managed configuration for the Apollo gateway is exposed through [Google Cloud Storage](https://cloud.google.com/storage). For all API keys, the Apollo Graph Manager provisions a public file accessible via the hash of the API key used to gain access to the managed configuration files for the graph to which the API key belongs. In the event that managed configuration is inaccessible, due to an outage in Google's Cloud Storage service, the gateway will continue to serve the last-known configuration. In the event that that Apollo Graph Manager API is down, changes to managed configuration will be stalled, but the last-published managed configuration files will still be accessible via GCS.

Reviewer 3: Simplifications for readability (and remove `that that`)

```suggestion
The managed configuration for the Apollo gateway is exposed through [Google Cloud Storage](https://cloud.google.com/storage). For all API keys, the Apollo Graph Manager provisions a public file accessible via the hash of the API key. In the event that managed configuration is inaccessible due to an outage in Google's Cloud Storage service, the gateway will continue to serve the last-known configuration. In the event that Apollo Graph Manager API is down, changes to managed configuration will be stalled but the last-published configuration files will still be accessible via GCS.
```

### Using variants to control rollout

> With [managed federation](#managed-federation), you have the ability to control which version of your graph a fleet of Gateways are running with. For the majority of deployments, rolling over all of your Gateways to a new schema version is a good strategy, since changes should be checked to be backwards compatible using [Schema Validation](/platform/schema-validation/). However, changes at the gateway level may involve a variety of different updates, like changing how query plans are generated or transferring type ownership from one service to another. In the case that your infrastructure requires more advanced deployment strategies, we recommend using [graph variants](/platform/schema-registry/#registering-schemas-to-a-variant) to manage different fleets of Gateways running with different configurations.

Reviewer 2: okay i take things back, i think we should say managed configuration here because it's being used in a lot of places. i think its best if we stick to just one.

> With [managed federation](#managed-federation), you have the ability to control which version of your graph a fleet of Gateways are running with. For the majority of deployments, rolling over all of your Gateways to a new schema version is a good strategy, since changes should be checked to be backwards compatible using [Schema Validation](/platform/schema-validation/). However, changes at the gateway level may involve a variety of different updates, like changing how query plans are generated or transferring type ownership from one service to another. In the case that your infrastructure requires more advanced deployment strategies, we recommend using [graph variants](/platform/schema-registry/#registering-schemas-to-a-variant) to manage different fleets of Gateways running with different configurations.

Reviewer 3: ```suggestion
With [managed federation](#managed-federation), you have the ability to control which version of your graph a fleet of gateways are running with. For the majority of deployments, rolling over all of your gateways to a new schema version is a good strategy, since changes should be checked to be backwards compatible using [Schema Validation](/platform/schema-validation/). However, changes at the gateway level may involve a variety of different updates, like transferring type ownership from one service to another. In the case that your infrastructure requires more advanced deployment strategies, we recommend using [graph variants](/platform/schema-registry/#registering-schemas-to-a-variant) to manage different fleets of gateways running with different configurations.
```
Gateway(s) -> gateway(s)

Removed - technically correct, but maybe unnecessary and seems to imply that this is configurable
>changing how query plans are generated or

For instance, in order to have a canary deployment, you might maintain two production graphs in the [Schema Registry][], one called `prod` and one called `prod-canary`. Your deployment of a change to some implementing service named "foo" might look something like this:

1. Check changes in "foo" against `prod` and `prod-canary`:
   ```
   apollo service:check --tag=prod --serviceName=foo && apollo service:check --tag=prod-canary --serviceName=foo
   ```
1. Deploy changes to "foo" into your production environment (_Note: This will not roll out changes to the gateway yet_)
1. Roll over the `prod-canary` graph, containing one gateway container, using
   ```
   apollo service:push --tag=prod-canary --serviceName=foo
   ```

>    (_Note: If composition fails due to intermediate changes to the canary graph, new configuration will not be rolled out_)

Reviewer 3: Typically use a quote for asides like this
```suggestion
>Note: If composition fails due to intermediate changes to the canary graph, new configuration will not be rolled out
```

1. Wait for health checks to pass against the canary, watch dashboards, etc.
1. After the canary is stable, roll out the changes to the rest of production using
  ```
  apollo service:push --tag=prod --serviceName=foo
  ```

Because you can [tag metrics with variants](/platform/schema-registry/#associating-metrics-with-a-variant) as well, you can use [Apollo Graph Manager](https://engine.apollographql.com) to verify a canary's performance before rolling out changes to the rest of the graph. You can also use a similar strategy with variants to support a variety of other advanced deployment workflows, like blue/green deployments.

# Running Federation in Production
> With a focus on reliability and robustness

> There is a wealth of information around reliability in a distributed service-oriented architecture, and those best practices still certainly apply here. For instance, you should ensure that the resource needs of your services are well-understood, that services and the gateway only roll over to a new version after passing health checks, and you should strive to make all services stateful when possible. This document is meant for teams with the need for a federated GraphQL layer, to provide details of nuance in how federation works, how the gateway responds to changes, and best practices in operating federation in production. If you're running with federation in production, we'd love to [hear from you](mailto:federation@apollographql.com) and accept any [contribution](https://github.com/apollographql/apollo-server/blob/master/docs/source/federation/production.md) to this documentation.

Reviewer 3: Added comma, removed hyphen, simplified a sentence

```suggestion
There is a wealth of information around reliability in a distributed, service-oriented architecture, and those best practices still certainly apply here. For instance, you should ensure that the resource needs of your services are well understood, that services and the gateway only roll over to a new version after passing health checks, and that all services are stateful when possible. This document is meant to provide details into how federation works, how the gateway responds to changes, and best practices in operating federation in production. If you're running with federation in production, we'd love to [hear from you](mailto:federation@apollographql.com) and accept any [contribution](https://github.com/apollographql/apollo-server/blob/master/docs/source/federation/production.md) to this documentation.
```

## Best Practices

In operating federation in production ourselves and working with a variety of teams deploying federation in their environments, we have collected some best practices to maintain reliability and control over a federated GraphQL layer at scale. If you're running federation in your infrastructure, we'd love to [hear from you](mailto:federation@apollographql.com) to help share any best practices you and your team may have learned from operating federation at scale.

### Treat the gateway as infrastructure

> The [Apollo Gateway](https://www.apollographql.com/docs/apollo-server/federation/implementing/#running-a-gateway) understands how to "speak" federation and orchestrate an incoming operation from a client into a set of operations to the underlying services that implement the graph. Because the gateway understands how to "speak" federation, it should contain **no business logic** and we recommend treating it as you would any infrastructure, such as a load balancer or a service discovery agent. There should be no need to re-deploy the gateway as services roll out underneath it. The gateway is designed to be able to smoothly roll over to updates in any services underneath it, and we recommend using the [managed federation](https://www.apollographql.com/docs/apollo-server/federation/production/#managed-federation) to control how changes in underlying services bubble up to the gateway. That said, like any infrastructure, you'll want to investigate the different [configuration options](https://www.apollographql.com/docs/apollo-server/api/apollo-gateway/) and choose the ones that make sense for your use case. Additionally, make sure to expose your gateway to sufficient load to understand its resource consumption and provision accordingly.

Reviewer 3: Instead of "speak" federation, can we say:
> The Apollo Gateway orchestrates requests to implementing services in order to fulfill client requests.

Reviewer 3: I don't understand how these are related via 'because x, y' :
>Because the gateway understands how to "speak" federation, it should contain **no business logic**

We may just prefer to lead with our stance:
>The gateway should be treated as a piece of infrastructure, containing little to no business logic...

### Put graph-level functionality in the gateway

> Because the gateway is the component of your GraphQL layer that sees and services every client operation, there are a number of things that make sense to happen at the gateway level. This includes functionality like [reporting metrics](https://www.apollographql.com/docs/apollo-server/features/metrics/), [whole response caching](https://www.apollographql.com/docs/apollo-server/features/caching/#saving-full-responses-to-a-cache), [operation safe-listing](/platform/operation-registry/), and [automatic persisted queries](https://www.apollographql.com/docs/apollo-server/features/apq/). Because all of this functionality happens at the operation-level, we recommend ensuring that the gateway is configured to support this functionality rather than exposing it piecemeal in each implementing service. Since the Apollo gateway is built atop Apollo Server, setting these features up should require nothing more than following the associated guides.

Reviewer 3: Remove hyphen, fix capitalization

```suggestion
Because the gateway is the component of your GraphQL layer that sees and services every client operation, there are a number of things that make sense to happen at the gateway level. This includes functionality like [reporting metrics](https://www.apollographql.com/docs/apollo-server/features/metrics/), [whole response caching](https://www.apollographql.com/docs/apollo-server/features/caching/#saving-full-responses-to-a-cache), [operation safe-listing](/platform/operation-registry/), and [automatic persisted queries](https://www.apollographql.com/docs/apollo-server/features/apq/). Because all of this functionality happens at the operation level, we recommend ensuring that the gateway is configured to support this functionality rather than exposing it piecemeal in each implementing service. Since the Apollo Gateway is built atop Apollo Server, setting these features up should require nothing more than following the associated guides.
```

Traces will be reported in the shape of the query plan, with each unique fetch to an implementing service reporting timing and error data.

<div style="text-align:center">
    <img src="../images/federated_trace.png" alt="Federated Trace">
    </img>
</div>

Operation-level statistics will still be collected over the operations sent by the client, and those operations will be validated as part of the `service:check` validation workflow.

### Inspecting Query Plans

<!--
add links to advanced-features for
- [`@requires`](/federation/advanced-features/#computed-fields)`
-`[`@provides`](/federation/advanced-features/#using-denormalized-data)`
 -->

> When the gateway receives a new query, it generates a query plan that defines the sequence of requests the gateway will send to the necessary downstream services. Inspecting a query plan can be a helpful tool in understanding the gateway and exploring how directives like `@requires` and `@provides` can help optimize query plans. To make it easy to access query plans, the `@apollo/gateway` package includes a build of GraphQL Playground that adds a query plan inspector.

Reviewer 3: Updated to say that AS includes our version of playground rather than the gateway package

```suggestion
When the gateway receives a new query, it generates a query plan that defines the sequence of requests the gateway will send to the necessary downstream services. Inspecting a query plan can be a helpful tool in understanding the gateway and exploring how directives like `@requires` and `@provides` can help optimize query plans. To make it easy to access query plans, Apollo Server includes a build of GraphQL Playground that adds a query plan inspector.
```

... More overview stuff

## Inspecting your Graph
> Using the `apollo service:list` command or the UI to understand your graph

To understand the implementing services that make up your graph, you can use both the Apollo command line interface as well as the Apollo Platform.

Use `apollo service:list` in the command line to see a snapshot of the services that make up your graph, including their endpoints and when they were last updated.

Here's what running `apollo service:list` will look like:

```
~$ apollo service:list
  ✔ Loading Apollo Project
  ✔ Fetching list of services for graph service-list-federation-demo

name       URL                            last updated

> ─────────  ─────────────────────────────  ────────────────────────

Reviewer 4: the labels above are still in a [PR](https://github.com/apollographql/apollo-tooling/pull/1377/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1e) but I realized they are necessary when writing this documentation:

Accounts   http://localhost:4001/graphql  3 July 2019 (2 days ago)

> Inventory  http://localhost:4004/graphql  3 July 2019 (2 days ago)

Reviewer 2: Do we order the service lists in any way? Like are they ordered by last updated? If they are it might cool to have different last updated entries so readers can see which ones were updated recently

Reviewer 4: They are alphabetical in the UI, which we can do here. I think we should keep it consistent between UI and CLI, no?
